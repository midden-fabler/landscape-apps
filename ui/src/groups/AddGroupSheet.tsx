import { useCallback, useState } from 'react';
import cn from 'classnames';
import WidgetDrawer from '@/components/WidgetDrawer';
import HomeIconMobileNav from '@/components/icons/HomeIconMobileNav';
import NewRaysIcon from '@/components/icons/NewRaysIcon';
import CaretLeft16Icon from '@/components/icons/CaretLeft16Icon';
import { strToSym } from '@/logic/utils';
import { useCreateGroupMutation } from '@/state/groups';
import { useNavigate } from 'react-router-dom';
import { isNativeApp } from '@/logic/native';

function DragHandle() {
  return (
    <div className="my-3 flex w-full justify-center">
      <div className="h-[5px] w-[32px] rounded-[100px] bg-gray-100" />
    </div>
  );
}

function CreateGroup(props: { back: () => void }) {
  const { mutateAsync: createGroupMutation, isLoading } =
    useCreateGroupMutation();
  const navigate = useNavigate();
  const [input, setInput] = useState('');
  const shortCode = strToSym(input).replace(/[^a-z]*([a-z][-\w\d]+)/i, '$1');

  const createGroup = useCallback(async () => {
    if (!input || !shortCode) return;

    try {
      await createGroupMutation({
        title: input,
        description: '',
        image: '',
        cover: '',
        name: shortCode,
        members: {},
        cordon: {
          open: {
            ships: [],
            ranks: [],
          },
        },
        secret: false,
      });

      const flag = `${window.our}/${shortCode}`;
      navigate(`/groups/${flag}`);
    } catch (error) {
      console.log("Couldn't create group", error);
    }
  }, [createGroupMutation, input, shortCode, navigate]);

  return (
    <div className="flex w-full flex-col items-center">
      <div className="mb-4 flex w-full items-center justify-between px-2">
        <div
          className="flex h-6 w-6 items-center justify-center"
          onClick={() => props.back()}
        >
          <CaretLeft16Icon className="h-6 w-6" />
        </div>
        <h3 className="text-[17px]">New Group</h3>
        <div className="invisible h-6 w-6" />
      </div>
      <div className="flex w-full flex-col pt-6">
        <label className="text-small pb-3 text-gray-400">
          Name your group, you can edit it later
        </label>
        <input
          className="w-full rounded-lg border border-gray-100 px-4 py-3.5 text-lg outline-none"
          autoFocus
          placeholder="Group name"
          value={input}
          onChange={(e) => setInput(e.target.value)}
        />
        <p className="text-small pt-4 text-gray-400">
          Your <span className="text-black">public</span> group will live at:
          <br />
          {window.ship || '~latter-bolden'}/group/
          <span className="text-black">
            {input !== '' ? shortCode : 'group-name'}
          </span>
        </p>
        <p className="text-small pt-6 text-gray-400">
          {' '}
          You can edit your group's privacy later.{' '}
        </p>
      </div>
      <button
        className="mt-6 w-full rounded-lg border border-blue-200 bg-blue-soft px-6 py-4 text-lg font-semibold text-blue disabled:border-gray-100 disabled:bg-gray-50 disabled:text-gray-400"
        disabled={input === ''}
        onClick={createGroup}
      >
        Create Group
      </button>
    </div>
  );
}

export default function AddGroupSheet(props: {
  open: boolean;
  onOpenChange: (open: boolean) => void;
}) {
  const [view, setView] = useState<'root' | 'create' | 'join'>('root');
  const height = view === 'root' ? 'h-30vh' : 'h-60vh';

  const onOpenChange = (open: boolean) => {
    if (!open) {
      setView('root');
    }
    props.onOpenChange(open);
  };

  const CreateOrJoin = useCallback(() => {
    return (
      <div className="flex w-full flex-col items-center">
        <h3 className="mt-4 mb-6 text-[17px] ">Add a group</h3>
        <div className="flex flex-col rounded-lg border border-gray-100">
          <button
            className="flex items-center border-b border-gray-100 px-6 py-4"
            onClick={() => setView('join')}
          >
            <HomeIconMobileNav className="mr-4 h-6 w-6" isInactive asIcon />
            <div className="flex flex-col items-start">
              <h4 className="mb-1.5 text-[17px] tracking-wide">Join a group</h4>
              <p className="text-gray-300">
                Join with short code or host's Urbit ID
              </p>
            </div>
          </button>
          <button
            className="flex items-center px-6 py-4 text-blue"
            onClick={() => setView('create')}
          >
            <NewRaysIcon className="mr-4 h-6 w-6" />
            <div className="flex flex-col items-start">
              <h4 className="mb-1.5 text-[17px] tracking-wide">
                Create new group
              </h4>
              <p className="text-gray-300">Start a group from scratch</p>
            </div>
          </button>
        </div>
      </div>
    );
  }, []);

  return (
    <WidgetDrawer
      open={props.open}
      onOpenChange={onOpenChange}
      className={cn('px-6', height)}
    >
      <DragHandle />
      <div className="mt-4">
        {view === 'root' && <CreateOrJoin />}
        {view === 'create' && <CreateGroup back={() => setView('root')} />}
      </div>
      {!isNativeApp() && <div className="pb-6" />}
    </WidgetDrawer>
  );
}
