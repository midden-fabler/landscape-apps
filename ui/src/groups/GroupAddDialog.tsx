import { Link, useLocation } from 'react-router-dom';
import WidgetDrawer from '@/components/WidgetDrawer';
import { useDismissNavigate } from '@/logic/routing';
import HomeIconMobileNav from '@/components/icons/HomeIconMobileNav';
import BangIcon from '@/components/icons/BangIcon';

export default function GroupAddDialog() {
  const dismiss = useDismissNavigate();
  const location = useLocation();
  return (
    <WidgetDrawer
      onOpenChange={(o) => !o && dismiss()}
      className="px-6 py-12"
      open
    >
      <h1 className="text-center text-[17px] font-medium">Add a group</h1>
      <ul className="mt-6 divide-y divide-gray-100 overflow-hidden rounded-xl border border-gray-100">
        <li>
          <Link
            to="/groups/join"
            state={location.state}
            className="flex items-center gap-4 py-4 px-6 active:bg-gray-100"
          >
            <HomeIconMobileNav className="h-6 w-6" isInactive />
            <div className="space-y-1">
              <h2 className="text-[17px] font-medium">Join a group</h2>
              <p className="text-gray-400">
                Join with short code or host&apos;s Urbit ID
              </p>
            </div>
          </Link>
        </li>
        <li>
          <Link
            to="/groups/create"
            state={location.state}
            className="flex items-center gap-4 py-4 px-6 text-blue active:bg-gray-100"
          >
            <BangIcon className="h-6 w-6" />
            <div className="space-y-1">
              <h2 className="text-[17px] font-medium">Create new group</h2>
              <p className="text-gray-400">Start a group from scratch</p>
            </div>
          </Link>
        </li>
      </ul>
    </WidgetDrawer>
  );
}
