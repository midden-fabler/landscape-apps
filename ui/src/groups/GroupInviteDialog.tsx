import ob from 'urbit-ob';
import React, { useCallback, useState } from 'react';
import cn from 'classnames';
import Dialog, { DialogClose } from '@/components/Dialog';
import ShipSelector, { ShipOption } from '@/components/ShipSelector';
import { useDismissNavigate } from '@/logic/routing';
import {
  useGroup,
  useGroupAddMembersMutation,
  useGroupInviteMutation,
  useRouteGroup,
} from '@/state/groups/groups';
import { useIsMobile } from '@/logic/useMedia';
import { getFlagParts, getPrivacyFromGroup, preSig } from '@/logic/utils';
import Sheet, { SheetContent } from '@/components/Sheet';
import LoadingSpinner from '@/components/LoadingSpinner/LoadingSpinner';
import ExclamationPoint from '@/components/icons/ExclamationPoint';
import HostConnection from '@/channels/HostConnection';
import { useConnectivityCheck } from '@/state/vitals';
import LureInviteBlock from './LureInviteBlock';

export function GroupInviteBlock() {
  const flag = useRouteGroup();
  const group = useGroup(flag);
  const host = getFlagParts(flag).ship;
  const { data } = useConnectivityCheck(host);
  const privacy = group ? getPrivacyFromGroup(group) : 'public';
  const [ships, setShips] = useState<ShipOption[]>([]);
  const validShips =
    ships && ships.length > 0
      ? ships.every((ship) => ob.isValidPatp(preSig(ship.value)))
      : false;
  const {
    mutate: inviteMutation,
    status: inviteStatus,
    reset: resetInvite,
  } = useGroupInviteMutation();
  const {
    mutate: addMembersMutation,
    status: addMembersStatus,
    reset: resetAddMembers,
  } = useGroupAddMembersMutation();
  const saga = group?.saga || null;
  const hasIssue =
    (saga !== null && !('synced' in saga)) ||
    (data?.status &&
      'complete' in data.status &&
      data.status.complete !== 'yes');

  const onInvite = useCallback(async () => {
    const shipList = ships.map((s) => preSig(s.value));

    try {
      if (privacy !== 'public') {
        inviteMutation({ flag, ships: shipList });
      } else {
        addMembersMutation({ flag, ships: shipList });
      }
      setShips([]);
      setTimeout(() => {
        resetInvite();
        resetAddMembers();
      }, 3000);
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error('Error inviting/adding members: poke failed');
      setTimeout(() => {
        resetInvite();
        resetAddMembers();
      }, 3000);
    }
  }, [
    flag,
    privacy,
    ships,
    inviteMutation,
    addMembersMutation,
    resetInvite,
    resetAddMembers,
  ]);

  return (
    <div className="card">
      <h2 className="mb-1 text-lg font-bold">Invite by Urbit ID</h2>
      <p className="mb-4 text-gray-600">
        (e.g. ~sampel-palnet) or display name
      </p>
      {hasIssue && (
        <HostConnection
          type="combo"
          className="mb-4 text-sm"
          ship={host}
          status={data?.status}
          saga={group?.saga || null}
        />
      )}
      <div className="w-full py-3">
        <ShipSelector
          ships={ships}
          setShips={setShips}
          onEnter={onInvite}
          placeholder="Search"
          autoFocus={false}
        />
      </div>
      <div className="flex items-center justify-end space-x-2">
        <DialogClose className="secondary-button">Cancel</DialogClose>
        {addMembersStatus === 'success' ? (
          <button disabled className="button">
            Invites Sent
          </button>
        ) : (
          <button
            onClick={onInvite}
            className={cn('button', {
              'bg-red':
                inviteStatus === 'error' || addMembersStatus === 'error',
            })}
            disabled={
              !validShips ||
              inviteStatus === 'loading' ||
              inviteStatus === 'error' ||
              addMembersStatus === 'loading' ||
              addMembersStatus === 'error'
            }
          >
            Send Invites
            {inviteStatus === 'loading' || addMembersStatus === 'loading' ? (
              <LoadingSpinner className="ml-2 h-4 w-4" />
            ) : null}
            {inviteStatus === 'error' || addMembersStatus === 'error' ? (
              <ExclamationPoint className="ml-2 h-4 w-4" />
            ) : null}
          </button>
        )}
      </div>
    </div>
  );
}

export default function GroupInviteDialog() {
  const dismiss = useDismissNavigate();
  const flag = useRouteGroup();
  const group = useGroup(flag);
  const isMobile = useIsMobile();

  function renderContent() {
    return (
      <div className={cn('flex flex-col space-y-6', !isMobile && 'mt-10')}>
        {group && (
          <>
            <LureInviteBlock flag={flag} group={group} />
            <GroupInviteBlock />
          </>
        )}
      </div>
    );
  }

  return isMobile ? (
    <Sheet open={true} onOpenChange={(o) => !o && dismiss()}>
      <SheetContent className="flex flex-col overflow-y-auto" showClose={false}>
        {renderContent()}
      </SheetContent>
    </Sheet>
  ) : (
    <Dialog
      open={true}
      onOpenChange={(isOpen) => !isOpen && dismiss()}
      containerClass="w-full max-w-xl"
      className="mb-64 bg-transparent p-0"
      close="none"
    >
      {renderContent()}
    </Dialog>
  );
}
