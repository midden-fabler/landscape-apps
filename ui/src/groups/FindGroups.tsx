import React, { useEffect, useState } from 'react';
import { Helmet } from 'react-helmet';
import { useNavigate, useParams, useLocation } from 'react-router';
import cn from 'classnames';
import ob from 'urbit-ob';
import {
  useGangs,
  useGroupIndex,
  usePendingGangsWithoutClaim,
} from '@/state/groups';
import { useIsMobile } from '@/logic/useMedia';
import ShipSelector, { ShipOption } from '@/components/ShipSelector';
import { Gangs, ViewProps } from '@/types/groups';
import { hasKeys, preSig, whomIsFlag } from '@/logic/utils';
import { useModalNavigate } from '@/logic/routing';
import GroupReference from '@/components/References/GroupReference';
import ReconnectingSpinner from '@/components/ReconnectingSpinner';
import ShipConnection from '@/components/ShipConnection';
import { useConnectivityCheck } from '@/state/vitals';
import GroupJoinList from './GroupJoinList';
import GroupJoinListPlaceholder from './GroupJoinListPlaceholder';
import GroupAvatar from './GroupAvatar';

export default function FindGroups({ title }: ViewProps) {
  const { ship, name } = useParams<{ ship: string; name: string }>();
  const location = useLocation();
  const navigate = useNavigate();
  const modalNavigate = useModalNavigate();
  const existingGangs = useGangs();
  const pendingGangs = usePendingGangsWithoutClaim();
  const isMobile = useIsMobile();
  const { groupIndex, fetchStatus, refetch } = useGroupIndex(ship || '');

  /**
   *  Search results for render:
   *
   *  - Filter out index results for gangs with invites, since they will be
   *    displayed in the Pending Invites section below.
   *  - Only show Public and Private groups, unless already a member of a ship's
   *    Secret group
   * */
  const indexedGangs = groupIndex
    ? Object.entries(groupIndex)
        .filter(([flag, preview]) => {
          if (flag in existingGangs) {
            // Bucket invites separately (in Pending Invites section)
            if (existingGangs[flag].invite) {
              return false;
            }
            // Show group if already a member
            return true;
          }

          // Hide secret gangs
          return !('afar' in preview.cordon);
        })
        .reduce(
          (memo, [flag, preview]) => ({
            ...memo,
            [flag]: {
              preview,
              invite: null,
              claim: flag in existingGangs ? existingGangs[flag].claim : null,
            },
          }),
          {} as Gangs
        )
    : null;

  const flag = name ? preSig(`${ship}/${name}`) : ship ? preSig(ship) : '';
  const gangToDisplay = name ? indexedGangs?.[flag] : null;
  // If needed group found - show it, otherwise show all groups
  const gangsToDisplay = gangToDisplay
    ? { [flag]: gangToDisplay }
    : indexedGangs;
  const hasResults = gangsToDisplay && hasKeys(gangsToDisplay);

  const [shipSelectorShips, setShipSelectorShips] = useState<ShipOption[]>([]);

  const selectedShip =
    shipSelectorShips.length > 0 ? shipSelectorShips[0] : null;
  let presentedShip = selectedShip
    ? selectedShip.label || selectedShip.value
    : '';

  const { data, showConnection } = useConnectivityCheck(ship || presentedShip);
  // Show only the name of the ship if the needed group is not in the results
  if (name && gangsToDisplay && !gangToDisplay) {
    [presentedShip] = presentedShip.split('/');
  }

  // once a ship is selected, redirect to find/[selected query]
  useEffect(() => {
    if (selectedShip) {
      navigate(`/find/${selectedShip.value}`);
      return;
    }

    if (ship && name) {
      navigate(`/find/${ship}/${name}`);
      return;
    }

    // user has cleared selection, redirect back to find root
    navigate('/find');
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selectedShip]);

  const handleProfileClick = () => {
    modalNavigate(`/profile/${preSig(ship!)}`, {
      state: { backgroundLocation: location },
    });
  };

  const resultsTitle = () => {
    if (fetchStatus === 'fetching') {
      return (
        <>
          <div className="mb-3 inline-block rounded bg-gray-50 p-2">
            <ShipConnection
              type="combo"
              ship={ship || presentedShip}
              status={data?.status}
            />
          </div>
          <p>
            Searching for groups hosted by&nbsp;
            <span
              onClick={handleProfileClick}
              className="cursor-pointer text-gray-800"
            >
              {presentedShip === '' ? ship : presentedShip}
            </span>
            <span>&nbsp;...</span>
          </p>
        </>
      );
    }

    if (hasResults) {
      return (
        <>
          <span>Groups hosted by&nbsp;</span>
          <span
            onClick={handleProfileClick}
            className="cursor-pointer text-gray-800"
          >
            {presentedShip === '' ? ship : presentedShip}
          </span>
          <span>:</span>
        </>
      );
    }

    return (
      <span>
        <div className="mb-3 inline-block rounded bg-gray-50 p-2">
          <ShipConnection
            type="combo"
            ship={ship || presentedShip}
            status={data?.status}
          />
        </div>
        <p>
          Your search timed out, which may happen when a ship hosts no groups,
          is under heavy load, or is offline.{' '}
          <span
            onClick={() => refetch()}
            className="cursor-pointer text-gray-800"
          >
            Try again?
          </span>
        </p>
      </span>
    );
  };

  // Allow selecting a ship name or invite URL (i.e., flag) in ShipSelector
  const isValidNewOption = (val: string) =>
    val ? ob.isValidPatp(preSig(val)) || whomIsFlag(val) : false;

  return (
    <>
      {isMobile && (
        <header className="flex items-center justify-between bg-white px-6 py-4 sm:hidden">
          <h1 className="text-lg font-bold text-gray-800">Find Groups</h1>
          <div className="flex shrink-0 flex-row items-center space-x-3 self-end">
            {isMobile && <ReconnectingSpinner />}
          </div>
        </header>
      )}
      <div className={cn('flex grow overflow-y-auto bg-gray-50')}>
        <Helmet>
          <title>{title ? title : document.title}</title>
        </Helmet>
        <div className="w-full p-6">
          {hasKeys(pendingGangs) ? (
            <section className={cn('card mb-6 space-y-4')}>
              <h1
                className={cn('font-bold', isMobile ? 'text-base' : 'text-lg')}
              >
                Pending Invites
              </h1>
              <GroupJoinList gangs={pendingGangs} />
            </section>
          ) : null}
          {!selectedShip && (
            <section className={cn('card mb-6')}>
              <h1 className="text-lg font-bold">Some Good Groups</h1>
              <p className="mb-2 mt-4 leading-6 text-gray-600">
                Here are some groups we like.
              </p>
              <p className="mt-2 mb-8 leading-6 text-gray-600">
                You can find many more in the Urbit Foundationâ€™s{' '}
                <a
                  href="https://urbit.org/ecosystem?type=groups"
                  target="_blank"
                  rel="noreferrer"
                >
                  directory
                </a>
                .
              </p>
              <div className="grid grid-cols-1 gap-4">
                <div className="flex items-center justify-between">
                  <GroupAvatar
                    image="https://interstellar.nyc3.digitaloceanspaces.com/battus-datsun/2022.11.07..19.39.22-Sig.png"
                    size="h-12 w-12 shrink-0"
                  />
                  <div className="mx-2 grow space-y-2">
                    <h2 className="text-base font-semibold">
                      Urbit Foundation
                    </h2>
                    <p className="text-gray-400">
                      Learn about the Urbit project
                    </p>
                  </div>
                  <GroupReference flag="~halbex-palheb/uf-public" onlyButton />
                </div>
                <div className="flex items-center justify-between">
                  <GroupAvatar
                    image="https://www.door.link/logowhite.svg"
                    size="h-12 w-12 shrink-0"
                  />
                  <div className="mx-2 grow space-y-2">
                    <h2 className="text-base font-semibold">door.link</h2>
                    <p className="text-gray-400">A cult of music lovers</p>
                  </div>
                  <GroupReference flag="~natnex-ronret/door-link" onlyButton />
                </div>
                <div className="flex items-center justify-between">
                  <GroupAvatar
                    image="https://nyc3.digitaloceanspaces.com/fabled-faster/fabled-faster/2023.4.06..02.45.53-tlon-local-pin.svg"
                    size="h-12 w-12 shrink-0"
                  />
                  <div className="mx-2 grow space-y-2">
                    <h2 className="text-base font-semibold">Tlon Local</h2>
                    <p className="text-gray-400">
                      Updates, announcements, and broadcasts from Tlon.
                    </p>
                  </div>
                  <GroupReference flag="~nibset-napwyn/tlon" onlyButton />
                </div>
              </div>
            </section>
          )}
          <section className={cn('card mb-6 space-y-8')}>
            <div>
              <h1 className="text-lg font-bold">Join With Code</h1>
              <p className="mt-4 mb-8 leading-6 text-gray-600">
                If you know the{' '}
                <abbr title="~sampel-palnet" className="cursor-help">
                  host ID
                </abbr>{' '}
                or{' '}
                <abbr title="~sampel-palnet/group-name" className="cursor-help">
                  shortcode
                </abbr>{' '}
                of a public group, you can enter it here to join.
              </p>
              <label htmlFor="flag" className="mb-2 block font-semibold">
                Host ID or Shortcode
              </label>
              <div className="flex flex-col space-y-2">
                <ShipSelector
                  ships={shipSelectorShips}
                  setShips={setShipSelectorShips}
                  isMulti={false}
                  isClearable={true}
                  isLoading={fetchStatus === 'fetching'}
                  hasPrompt={false}
                  placeholder={'e.g. ~nibset-napwyn/tlon'}
                  isValidNewOption={isValidNewOption}
                  autoFocus={isMobile ? false : true}
                />
              </div>
            </div>
            {selectedShip || (ship && name) ? (
              <section className="space-y-3">
                <p className="font-semibold text-gray-400">{resultsTitle()}</p>
                {fetchStatus === 'fetching' ? (
                  <GroupJoinListPlaceholder />
                ) : hasResults ? (
                  <GroupJoinList gangs={gangsToDisplay} />
                ) : null}
              </section>
            ) : null}
          </section>
        </div>
      </div>
    </>
  );
}
