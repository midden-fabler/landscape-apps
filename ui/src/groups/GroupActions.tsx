import cn from 'classnames';
import React, { PropsWithChildren, useCallback, useState } from 'react';
import { Link, useLocation } from 'react-router-dom';
import EllipsisIcon from '@/components/icons/EllipsisIcon';
import { useChatState, usePinnedGroups } from '@/state/chat';
import useIsGroupUnread from '@/logic/useIsGroupUnread';
import UnreadIndicator from '@/components/Sidebar/UnreadIndicator';
import { citeToPath, getPrivacyFromGroup, useCopy } from '@/logic/utils';
import {
  useAmAdmin,
  useGang,
  useGroup,
  useGroupCancelMutation,
} from '@/state/groups';
import ActionMenu, { Action } from '@/components/ActionMenu';
import { Saga } from '@/types/groups';
import { ConnectionStatus } from '@/state/vitals';
import HostConnection from '@/channels/HostConnection';
import { useIsMobile } from '@/logic/useMedia';

const { ship } = window;

export function useGroupActions(flag: string) {
  const { doCopy } = useCopy(citeToPath({ group: flag }));
  const [isOpen, setIsOpen] = useState(false);
  const [copyItemText, setCopyItemText] = useState('Copy Group Link');
  const pinned = usePinnedGroups();
  const isPinned = Object.keys(pinned).includes(flag);

  const onCopy = useCallback(() => {
    doCopy();
    setCopyItemText('Copied!');
    setTimeout(() => {
      setCopyItemText('Copy Group Link');
      setIsOpen(false);
    }, 2000);
  }, [doCopy]);

  const onPinClick = useCallback(
    (e: React.MouseEvent) => {
      e.stopPropagation();
      useChatState.getState().togglePin(flag, !isPinned);
    },
    [flag, isPinned]
  );

  return {
    isOpen,
    isPinned,
    setIsOpen,
    copyItemText,
    onCopy,
    onPinClick,
  };
}

type GroupActionsProps = PropsWithChildren<{
  flag: string;
  saga?: Saga | null;
  status?: ConnectionStatus;
  className?: string;
}>;

const GroupActions = React.memo(
  ({ flag, saga, status, className, children }: GroupActionsProps) => {
    const { isGroupUnread } = useIsGroupUnread();
    const { claim } = useGang(flag);
    const location = useLocation();
    const hasActivity = isGroupUnread(flag);
    const group = useGroup(flag);
    const privacy = group ? getPrivacyFromGroup(group) : 'public';
    const isAdmin = useAmAdmin(flag);
    const isMobile = useIsMobile();
    const { mutate: cancelJoinMutation } = useGroupCancelMutation();
    const { isOpen, setIsOpen, isPinned, copyItemText, onCopy, onPinClick } =
      useGroupActions(flag);

    const onCopySelect = useCallback(
      (event: React.MouseEvent) => {
        event.preventDefault();
        onCopy();
      },
      [onCopy]
    );

    const actions: Action[] = [];

    if (saga && isMobile) {
      actions.push({
        key: 'connectivity',
        keepOpenOnClick: true,
        content: (
          <HostConnection
            ship={flag}
            status={status}
            saga={saga}
            type="combo"
            className="-ml-1 text-[17px] font-medium text-gray-800"
          />
        ),
      });
    }

    if (privacy === 'public' || isAdmin) {
      actions.push({
        key: 'invite',
        type: 'prominent',
        content: (
          <Link
            to={`/groups/${flag}/invite`}
            state={{ backgroundLocation: location }}
          >
            Invite People
          </Link>
        ),
      });
    }

    actions.push(
      {
        key: 'copy',
        onClick: onCopySelect,
        content: copyItemText,
        keepOpenOnClick: true,
      },
      {
        key: 'pin',
        onClick: onPinClick,
        content: isPinned ? 'Unpin' : 'Pin',
      },
      {
        key: 'settings',
        onClick: () => setIsOpen(false),
        content: isAdmin ? (
          <Link
            to={`/groups/${flag}/edit`}
            state={{ backgroundLocation: location }}
          >
            Group Settings
          </Link>
        ) : (
          <Link
            to={`/groups/${flag}/info`}
            state={{ backgroundLocation: location }}
          >
            Group Members & Info
          </Link>
        ),
      }
    );

    if (!flag.includes(ship)) {
      actions.push({
        key: 'leave',
        type: 'destructive',
        content: (
          <Link
            to={`/groups/${flag}/leave`}
            state={{ backgroundLocation: location }}
          >
            Leave Group
          </Link>
        ),
      });
    }

    if (claim) {
      actions.push({
        key: 'cancel_join',
        type: 'destructive',
        onClick: () => {
          cancelJoinMutation({ flag });
          setIsOpen(false);
        },
        content: 'Cancel Join',
      });
    }

    return (
      <ActionMenu
        open={isOpen}
        onOpenChange={setIsOpen}
        actions={actions}
        className={className}
      >
        {children || (
          <div className="relative h-6 w-6">
            {!isOpen && hasActivity ? (
              <UnreadIndicator
                className="absolute h-6 w-6 text-blue transition-opacity group-focus-within:opacity-0 group-hover:opacity-0"
                aria-label="Has Activity"
              />
            ) : null}
            <button
              className={cn(
                'default-focus absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 rounded-lg p-0.5 transition-opacity focus-within:opacity-100 hover:opacity-100 group-focus-within:opacity-100 group-hover:opacity-100',
                hasActivity && 'text-blue',
                isOpen ? 'opacity:100' : 'opacity-0'
              )}
              aria-label="Open Group Options"
            >
              <EllipsisIcon className="h-6 w-6" />
            </button>
          </div>
        )}
      </ActionMenu>
    );
  }
);

export default GroupActions;
