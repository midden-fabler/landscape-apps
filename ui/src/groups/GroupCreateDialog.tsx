import { useForm } from 'react-hook-form';

import { useNavigate } from 'react-router';
import WidgetDrawer from '@/components/WidgetDrawer';
import { useDismissNavigate } from '@/logic/routing';
import { useCreateGroupMutation } from '@/state/groups';
import { strToSym } from '@/logic/utils';

type GroupCreateSchema = {
  title: string;
};

const createGroupName = (title: string) =>
  strToSym(title).replace(/[^a-z]*([a-z][-\w\d]+)/i, '$1');

export default function GroupCreateDialog() {
  const navigate = useNavigate();
  const dismiss = useDismissNavigate();
  const { mutateAsync: createGroupMutation, status } = useCreateGroupMutation();
  const { register, watch, handleSubmit } = useForm<GroupCreateSchema>({
    defaultValues: {
      title: '',
    },
  });

  const groupTitle = watch('title');
  const groupSlug = `${window.our}/${
    createGroupName(groupTitle) || 'group-name'
  }`;

  const onSubmit = handleSubmit(async ({ title }) => {
    try {
      await createGroupMutation({
        title,
        name: createGroupName(title),
        description: '',
        image: '',
        cover: '',
        cordon: {
          shut: {
            pending: [],
            ask: [],
          },
        },
        members: {},
        secret: true,
      });
      navigate(`/groups/${groupSlug}`);
    } catch (err) {
      console.error('Error creating group:', err);
    }
  });

  return (
    <WidgetDrawer
      onOpenChange={(o) => !o && dismiss()}
      className="px-6 py-12"
      open
    >
      <h1 className="text-center text-[17px] font-medium">New Group</h1>
      <form onSubmit={onSubmit} className="mt-6 space-y-4 text-gray-400">
        <label htmlFor="title">Name your group, you can edit it later</label>
        <input
          {...register('title')}
          id="name"
          className="w-full rounded-lg border border-gray-200 bg-transparent p-4 text-[17px] font-medium"
          type="text"
          placeholder="Group name"
        />
        <p>
          Your <span className="text-gray-800">secret</span> group will live at:
          <br />
          {groupSlug}
        </p>
        <p>You can edit your group&apos;s privacy later.</p>
        <button
          className="w-full rounded-xl border border-blue/20 bg-blue-soft/90 px-6 py-5 text-left text-[17px] font-medium text-blue active:bg-blue-soft disabled:border-gray-400/50 disabled:bg-gray-50 disabled:text-gray-400"
          type="submit"
          disabled={!groupTitle || status === 'loading'}
        >
          Create Group
        </button>
      </form>
    </WidgetDrawer>
  );
}
