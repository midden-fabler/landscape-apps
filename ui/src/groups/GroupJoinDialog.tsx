import { useState } from 'react';
import ob from 'urbit-ob';
import ShipSelector, { ShipOption } from '@/components/ShipSelector';
import WidgetDrawer from '@/components/WidgetDrawer';
import { useDismissNavigate } from '@/logic/routing';
import { hasKeys, preSig, whomIsFlag } from '@/logic/utils';
import { useGangs, useGroupIndex } from '@/state/groups';
import type { Gangs } from '@/types/groups';
import GroupJoinList from './GroupJoinList';

export default function GroupJoinDialog() {
  const dismiss = useDismissNavigate();
  const [selectedShips, setSelectedShips] = useState<ShipOption[]>([]);
  const [ship, name] = selectedShips[0]?.value.split('/') ?? [];
  const { groupIndex, fetchStatus, refetch } = useGroupIndex(ship || '');
  const existingGangs = useGangs();

  /**
   *  Search results for render:
   *
   *  - Filter out index results for gangs with invites, since they will be
   *    displayed in the Pending Invites section below.
   *  - Only show Public and Private groups, unless already a member of a ship's
   *    Secret group
   * */
  const indexedGangs = groupIndex
    ? Object.entries(groupIndex)
        .filter(([flag, preview]) => {
          if (flag in existingGangs) {
            // Bucket invites separately (in Pending Invites section)
            if (existingGangs[flag].invite) {
              return false;
            }
            // Show group if already a member
            return true;
          }

          // Hide secret gangs
          return !('afar' in preview.cordon);
        })
        .reduce(
          (memo, [flag, preview]) => ({
            ...memo,
            [flag]: {
              preview,
              invite: null,
              claim: flag in existingGangs ? existingGangs[flag].claim : null,
            },
          }),
          {} as Gangs
        )
    : null;

  const flag = name ? preSig(`${ship}/${name}`) : ship ? preSig(ship) : '';
  const gangToDisplay = name ? indexedGangs?.[flag] : null;
  // If needed group found - show it, otherwise show all groups
  const gangsToDisplay = gangToDisplay
    ? { [flag]: gangToDisplay }
    : indexedGangs;
  const hasResults = gangsToDisplay && hasKeys(gangsToDisplay);

  // Allow selecting a ship name or invite URL (i.e., flag) in ShipSelector
  const isValidNewOption = (val: string) =>
    val ? ob.isValidPatp(preSig(val)) || whomIsFlag(val) : false;

  return (
    <WidgetDrawer
      onOpenChange={(o) => !o && dismiss()}
      className="px-6 py-12"
      open
    >
      <h1 className="text-center text-[17px] font-medium">Join a Group</h1>
      <div className="mt-6 space-y-4 text-gray-400">
        <label htmlFor="flag">
          Join existing groups with a host&apos;s name, Urbit ID, or a group
          invite shortcode.
        </label>
        <ShipSelector
          ships={selectedShips}
          setShips={setSelectedShips}
          isMulti={false}
          isClearable={true}
          isLoading={fetchStatus === 'fetching'}
          hasPrompt={false}
          placeholder="Host name or group shortcode"
          isValidNewOption={isValidNewOption}
        />
      </div>
      {hasResults ? (
        <div className="mt-4 max-h-[60vh] overflow-y-auto">
          <GroupJoinList gangs={gangsToDisplay} />
        </div>
      ) : null}
    </WidgetDrawer>
  );
}
