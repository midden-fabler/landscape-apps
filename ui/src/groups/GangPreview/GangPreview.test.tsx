import { describe, it, expect } from 'vitest';
import React from 'react';
import GangPreview from './GangPreview';
import { mockGangs } from '../../mocks/groups';
import { render } from '../../../test/utils';

describe('GangPreview', () => {
  it('renders preview', () => {
    const { asFragment } = render(
      <GangPreview preview={mockGangs['~zod/structure'].preview!} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
