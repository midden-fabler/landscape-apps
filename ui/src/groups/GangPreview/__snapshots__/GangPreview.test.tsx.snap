// Vitest Snapshot v1

exports[`GangPreview > renders preview 1`] = `
<DocumentFragment>
  <div
    class="flex space-x-4 rounded border p-2"
  >
    <div
      class="rounded h-16 w-16 text-xl flex items-center justify-center rounded-md"
      style="background-color: rgb(229, 229, 229);"
    >
      <span
        style="color: black;"
      >
        S
      </span>
    </div>
    <div
      class="flex flex-col justify-around"
    >
      <h4
        class="text-lg font-bold"
      >
        Structure
      </h4>
      <p>
        Urbit Structural Design and Engineering Group. Always Thinking About Mechanics.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
